### PR TITLE
Use systemd to manage service startup when available

### PR DESCRIPTION
**Pull request by maxbes**
When launching slapd through "slapd-cli start", on a systemd install, the process will not be seen as active by systemd (systemctl status slapd). In order to be correctly supervised, the slapd process must be started by systemd even when `slapd-cli start` is run.

This commit makes slapd startup/shutdown go through systemd, when
available (/run/systemd/system exists). 

Since systemd calls slapd-cli to start open ldap too, to avoid infinite
recursion, we need if slapd-cli is being run by a user or by systemd
itself (PPID=1)

Finally, SYSTEMCTL_SKIP_REDIRECT allows the behavior to be bypassed

Heavily inspired by legacy initscript behavior on RHEL7
(/etc/init.d/functions)

Tested on Debian 9 and Centos 7 so far

### CODE REVIEW

**Comment by coudot** [comment]
It does not seem to work if the LTB Debian package has installed the init script instead of systemd. But this cas should not happen on recent installations.

@davidcoutadeur for which version was done the switch to systemd? I have 2.4.47 installed on my machine, but still running with init.d, anf my version is buster/sid:
```
clement@ader-worteks:~/dev/worteks/openldap-rpm$ cat /etc/debian_version 
buster/sid
```

**Comment by coudot** [comment]
Another remark, the init script calls lsb init functions, which already makes a systemctl redirect:
```
root@ader-worteks:~# sh -x /etc/init.d/slapd status
+ VERSION=2.3
+ . /lib/lsb/init-functions
+ run-parts --lsbsysinit --list /lib/lsb/init-functions.d
+ [ -r /lib/lsb/init-functions.d/20-left-info-blocks ]
+ . /lib/lsb/init-functions.d/20-left-info-blocks
+ [ -r /lib/lsb/init-functions.d/40-systemd ]
+ . /lib/lsb/init-functions.d/40-systemd
+ _use_systemctl=0
+ [ -d /run/systemd/system ]
+ prog=slapd
+ service=slapd.service
+ systemctl -p LoadState --value show slapd.service
+ state=loaded
+ [ loaded = masked ]
+ [ 6188 -ne 1 ]
+ [ -z  ]
+ readlink -f /etc/init.d/slapd
+ [ loaded != not-found ]
+ _use_systemctl=1
+ systemctl -p CanReload --value show slapd.service
+ [ no = no ]
+ [ status = reload ]
+ [ 1 = 1 ]
+ set +e
+ set +u
+ [ xstatus = xstart -o xstatus = xstop -o xstatus = xrestart -o xstatus = xreload -o xstatus = xforce-reload -o xstatus = xtry-restart -o xstatus = xstatus ]
+ systemctl_redirect /etc/init.d/slapd status
+ local s
+ local rc
+ local prog=slapd
+ local command=status
+ service=slapd.service
+ systemctl is-system-running
+ OUT=running
+ [ status = status ]
+ /bin/systemctl --no-pager status slapd.service
● slapd.service - LSB: OpenLDAP
   Loaded: loaded (/etc/init.d/slapd; generated)
   Active: active (exited) since Fri 2019-08-23 12:08:39 CEST; 8min ago
     Docs: man:systemd-sysv-generator(8)
  Process: 6674 ExecStop=/etc/init.d/slapd stop (code=killed, signal=TERM)
  Process: 13086 ExecStart=/etc/init.d/slapd start (code=killed, signal=TERM)

août 23 12:07:08 ader-worteks systemd[1]: Starting LSB: OpenLDAP...
août 23 12:07:08 ader-worteks slapd[13086]: slapd-cli: [INFO] Using /usr/local/openldap/etc/openldap/slapd-cli.conf for configuration
août 23 12:07:08 ader-worteks slapd[13086]: slapd-cli: [INFO] Launching OpenLDAP configuration test...
août 23 12:07:08 ader-worteks slapd[13086]: slapd-cli: [OK] OpenLDAP configuration test successful
août 23 12:07:08 ader-worteks slapd[13086]: slapd-cli: [INFO] No db_recover done
août 23 12:07:08 ader-worteks slapd[13086]: slapd-cli: [INFO] Launching OpenLDAP...
août 23 12:07:08 ader-worteks slapd[13086]: slapd-cli: [OK] File descriptor limit set to 1024
août 23 12:08:39 ader-worteks systemd[1]: Started LSB: OpenLDAP.
+ rc=0
+ [ status = status ]
+ return 0
+ exit 0
```

**Comment by davidcoutadeur** [comment]
@coudot systemd is included since initscript v2.3, which is included since openldap-ltb 2.4.46

**Comment by maxbes** [comment]
I moved the systemd stuff after the sourcing of slapd-cli.conf and added another test to handle the case in which the initscript is still used. Can you retest?

**Comment by coudot** [comment]
Hello @maxbes it works well!

I can now use slapd-cli  and systemctl to manage slapd process, even if my installation relies on /etc/init.d/slapd

@davidcoutadeur I think this is an interesting improvement for our OpenLDAP packages. What do you think?

**Comment by davidcoutadeur** [comment]
Not sure what to think for now, nor the impact. Will give a look on this week-end. @coudot I don't think you planned to include this in 2.4.48 release, do you?

**Comment by coudot** [comment]
> I don't think you planned to include this in 2.4.48 release, do you?

I was thinking of it, yes. It will need to release initscript v2.5, but it should be quick.

**Comment by davidcoutadeur** [comment]
> 
> 
> > I don't think you planned to include this in 2.4.48 release, do you?
> 
> I was thinking of it, yes. It will need to release initscript v2.5, but it should be quick.

So you planned to rebuild rpm packages? Did you consider overwritting existing 2.4.48 or generate a new 2.4.48-2 release?

For debian package, I would have to:
- update changelog to include new initscript
- include new initscript and regenerate all packages
- update git repositories, create corresponding issues
- create a new release 2.4.48-2 or move the tag to one commit forward, and update github release

File: slapd-cli [hunk 1]
```diff
@@ -227,6 +227,17 @@ else
 	message "info" "[INFO] Using built-in configuration - this may cause some problems"
 fi
 
+# Use systemd to manage the service if it's present
+_use_systemctl=0
```
Comment (davidcoutadeur) [inline slapd-cli]: Maybe add comments to specify the signification of PPID, SYSTEMCTL_SKIP_REDIRECT, and INVOCATION_ID?

File: slapd-cli [hunk 2]
```diff
@@ -227,6 +227,17 @@ else
...
+if [ $PPID -ne 1 -a -z "$SYSTEMCTL_SKIP_REDIRECT" ] && \
+    [ -d /run/systemd/system ] && \
```
Comment (davidcoutadeur) [inline slapd-cli]: Not sure if  -d /run/systemd/system is a sufficient test, for example when slapd is run into docker.
See for example: https://github.com/ltb-project/openldap-deb/commit/b1c55b585e156b670fa5f85bfb5251f2dcf2d41c

```
  if ! type "systemctl" > /dev/null 2>&1; then
    # No SYSTEMD available
    # do not start on boot
    # do not start slapd process now
    echo && exit $?
  else
```

File: slapd-cli (hunk 1 above)
Comment (maxbes) [reply slapd-cli]: done

File: slapd-cli (hunk 2 above)
Comment (maxbes) [reply slapd-cli]: In a docker container, /run/systemd does not exist because systemd is not running in the container. In that case, slapd-cli will not redirect to systemd.

Testing whether systemctl exists or not is not a good test, because a user could have systemctl installed, but NOT be using systemd as their init system (corner case, maybe)

[ -d /run/systemd/system] is the test distros already do to see whether or not systemd is PID1 or not. See /lib/lsb/init-functions.d/40-systemd in debian and /etc/init.d/functions on redhat. So I am pretty confident about this being the right test to check whether systemd is currently managing the systemd or not.

**Comment by maxbes** [comment]
I have added some comments in code and hopefully clarified your other question.

However, I have discovered in the meantime that the INVOCATION_ID test that handles situations where 
* An initscript is present
* the /lib/systemd/system/slapd.service unit file is missing
* systemd is used as PID 1
... is not available in systemd <232 (debian jessie and ubuntu xenial)

I'm not sure exactly how you handle publication of debian packages, but I would recommend that you only publish these changes to stretch and later releases.

**Comment by davidcoutadeur** [comment]
> I'm not sure exactly how you handle publication of debian packages, but I would recommend that you only publish these changes to stretch and later releases.

Not so easy... The best would be to have a more generic initscript, because:
- it is where you make all the tests on systemd
- users could use their own compiled openldap, and download our initscript separately.

I will check this on my debian wheezy + jessie and try to find a solution

**Comment by maxbes** [comment]
A simple solution would be if /etc/init.d/slapd did a`exec ${CLI_BIN} [something]` instead of just `${CLI_BIN} something`, this way the test on PPID would work. But since /etc/init.d/slapd is a conffile, if won't be automatically replaced by package updates

**Comment by davidcoutadeur** [comment]
Tested just now with my debian environments:
- debian 7 32bits, debian 7 64bits, debian 9 64bits, debian 10 64bits : OK
- debian 8 64bits + systemd 215 : KO

Note that on debian 8 there is still an initscript deployed.

**Comment by maxbes** [comment]
I don't see any other environment variable we could use in jessie to break the loop. Why can't we just install the slapd.service file instead? It should work fine

**Comment by davidcoutadeur** [comment]
@maxbes: indeed, I didn't find any better solution. I suppose the easiest is to make an exec in the init file as you suggested.
Anyway, I discovered another problem in debian 9 + 10: when slapd is stopped, making a slapd-cli status produce some errors getting the PID: 

```
slapd-cli: [INFO] Using /usr/local/openldap/etc/openldap/slapd-cli.conf for configuration
slapd-cli: [INFO] LDAP Tool Box OpenLDAP init script version 2.4
cat: /usr/local/openldap/var/run/slapd.pid: Aucun fichier ou dossier de ce type
slapd-cli: [INFO] Process OpenLDAP is running (PID )
slapd-cli: [INFO] Listening to services ldap://*:389 ldaps://*:636 ldapi://%2Fvar%2Frun%2Fslapd%2Fldapi
error: process ID list syntax error

Usage:
 ps [options]

 Try 'ps --help <simple|list|output|threads|misc|all>'
  or 'ps --help <s|l|o|t|m|a>'
 for additional help text.

For more details see ps(1).
error: process ID list syntax error

Usage:
 ps [options]

 Try 'ps --help <simple|list|output|threads|misc|all>'
  or 'ps --help <s|l|o|t|m|a>'
 for additional help text.

For more details see ps(1).
slapd-cli: [INFO] Process usage: % CPU / % MEM
slapd-cli: [INFO] Detected suffix: dc=my-domain,dc=com

```

Will give a look on tomorrow.

**Comment by maxbes** [comment]
Hmm, his can only happen if slapd_status returns 0 (running) but the pidfile doesn't exist. I don't reproduce this unless I start the service then manually delete the pidfile. But I don't see how this can happen after stopping slapd.

**Comment by davidcoutadeur** [comment]
I can't reproduce the problem. As I did a lot of tests, maybe I was in an unstable state of systemd...
Anyway, it seems all good for me, and I'll publish the last fix for debian <= 8

@coudot: what about the release issue? If you intend to include initscript, we will need a new 2.5 initscript release too by the way.
> So you planned to rebuild rpm packages? Did you consider overwritting existing 2.4.48 or generate a new 2.4.48-2 release?
